File: Sample/Source/Scenes/SceneTitle.py
# -*- encoding: utf-8 -*-

from typing import Any, Dict
from Engine.UI import Image
from Engine.UI.Base import FunctionalBase
from Global import Manager, System, SceneBase
from Source import System as SourceSystem, System as GameSystem
from Source.Windows import WindowCommand, WindowSaveLoad
from Source.GameInstance import GameInstance


class Scene(SceneBase):
    r"""\brief Title screen scene with a command window for menu options."""

    def onEnter(self) -> None:
        r"""\brief Start with a blind transition effect."""
        System.setTransition(Manager.loadTransition("001-Blind01.png"), 3)

    def onCreate(self) -> None:
        r"""\brief Create background, command window, and load UI elements."""
        self._bg = Image(Manager.loadSystem("GrassBackground.png"))
        self._windowCommand = WindowCommand(
            ((0, 0), (320, 96)),
            {
                "Start": {"text": LOC("TITLE_START"), "callback": Scene._startGame},
                "Load": {"text": LOC("TITLE_CONTINUE"), "callback": lambda obj, kwargs: self._onLoadCommand()},
                "Config": {"text": LOC("TITLE_CONFIG"), "callback": Scene._configGame},
                "Exit": {"text": LOC("TITLE_EXIT"), "callback": Scene._exitGame},
            },
            columns=2,
        )
        self._windowCommand.setOrigin((160, 64))
        self._windowCommand.setPosition((320, 240))
        self._windowSaveLoad = WindowSaveLoad(
            slotRect=((112, 112), (160, 256)),
            detailRect=((272, 112), (256, 256)),
            loadOnly=True,
            onClose=self._onSaveLoadClose,
            onLoaded=self._onSaveLoadLoaded,
        )
        self._uiManager.loadUI(self._bg)
        self._uiManager.loadUI(self._windowCommand)
        self._uiManager.loadUI(self._windowSaveLoad.getSlotWindow())
        self._uiManager.loadUI(self._windowSaveLoad.getDetailWindow())
        self._titleBGM = None
        titleBGMFile = SourceSystem.getTitleBGM()
        if titleBGMFile:
            self._titleBGM = Manager.playMusic("BGM", titleBGMFile)
            if self._titleBGM is not None:
                self._titleBGM.setLooping(True)

    def onQuit(self) -> None:
        r"""\brief Stop title BGM when leaving this scene."""
        Manager.stopMusic("BGM")
        self._titleBGM = None

    def onDestroy(self) -> None:
        r"""\brief Ensure title BGM is stopped when scene is destroyed."""
        Manager.stopMusic("BGM")
        self._titleBGM = None

    @staticmethod
    def _startGame(obj: FunctionalBase, kwargs: Dict[str, Any]) -> None:
        from .SceneMap import Scene as SceneMap

        Manager.playSE(GameSystem.getDecisionSE())
        Manager.stopMusic("BGM")
        System.setScene(SceneMap())
        Cast(SceneMap, System.getScene()).setInst(GameInstance())

    def _onLoadCommand(self) -> None:
        Manager.playSE(GameSystem.getDecisionSE())
        self._windowCommand.setActive(False)
        self._windowSaveLoad.open()

    def _onSaveLoadClose(self, reason: str) -> None:
        if reason == "loaded":
            return
        self._windowCommand.setActive(True)

    def _onSaveLoadLoaded(self, inst: GameInstance) -> None:
        from .SceneMap import Scene as SceneMap

        Manager.stopMusic("BGM")
        nextScene = SceneMap()
        nextScene.setInst(inst)
        System.setScene(nextScene)

    @staticmethod
    def _configGame(obj: FunctionalBase, kwargs: Dict[str, Any]) -> None:
        Manager.playSE(GameSystem.getDecisionSE())
        print("Config Game")

    @staticmethod
    def _exitGame(obj: FunctionalBase, kwargs: Dict[str, Any]) -> None:
        Manager.playSE(GameSystem.getDecisionSE())
        System.exit()

File: Sample/Scripts/Source/Scenes/SceneTitle.lua
local Engine = require("Engine")
local GlobalCore = require("GlobalCore")
local GlobalFunctions = require("GlobalFunctions")
local SourceSystem = require("Source.System")
local GameInstance = require("Source.GameInstance")
local SceneTitleUI = require("Source.UI.Title")

local ManagerFunctions = GlobalFunctions.Manager
local AudioManager = GlobalCore.AudioManager
local GlobalSystem = GlobalCore.System
local SceneBase = GlobalCore.SceneBase

---@class Source.Scenes.SceneTitle: GlobalCore.SceneBase
local Scene = {}

---@diagnostic disable-next-line: unused
function Scene:onEnter()
    GlobalSystem.setTransition(ManagerFunctions.loadTransition("Flat.png"))
end

function Scene:onCreate()
    local WindowSaveLoad = require("Source.Windows.WindowSaveLoad")
    local ConfigWindow = require("Source.Windows.ConfigWindow")

    local uiManager = self:getUIManager()
    ---@cast uiManager GlobalCore.UIManager
    uiManager:setFocusNavigationEnabled(true)
    self._ui = SceneTitleUI.new(self)
    self._ui:mount(self:getUIManager(), GlobalSystem.getGameSize())
    self._windowCommand = self._ui:getCommandWindow()
    self._windowSaveLoad = WindowSaveLoad.new(
        nil, Engine.ToIntRect(112, 112, 160, 256), Engine.ToIntRect(272, 112, 256, 256), true, nil,
        function (reason)
            self:_onSaveLoadClose(reason)
        end,
        function (inst)
            self:_onSaveLoadLoaded(inst)
        end
    )
    self._configWindow = ConfigWindow.new(function ()
        self:_onConfigClose()
    end)
    ---@type any[]
    local uiWindows = { self._windowCommand, self._windowSaveLoad, self._configWindow }
    for _, window in ipairs(uiWindows) do
        ---@cast window Engine.ControlBase
        uiManager:loadUI(window)
    end
    self._windowCommand:setActive(false)
    self._ui:playAnimation("FadeIn", "CommandPanel", function ()
        self._ui:stopAnimation("FadeIn", "CommandPanel")
        self._windowCommand:setActive(true)
        self._windowCommand:requestKeyboardFocus()
    end)
    self._titleBGM = nil
    local titleBGMFile = SourceSystem.GetTitleBGM()
    if bool(titleBGMFile) then
        self._titleBGM = AudioManager.playMusic("BGM", titleBGMFile)
        if self._titleBGM ~= nil then
            self._titleBGM:setLooping(true)
        end
    end
end

function Scene:onQuit()
    ManagerFunctions.stopMusic("BGM")
    self._titleBGM = nil
end

function Scene:onDestroy()
    ManagerFunctions.stopMusic("BGM")
    self._titleBGM = nil
    self._windowSaveLoad:dispose()
    self._configWindow:dispose()
    self._ui:dispose()
end

---@diagnostic disable-next-line: unused
function Scene:_startGame()
    local SceneMap = require("Source.Scenes.SceneMap")

    AudioManager.playSound(SourceSystem.GetDecisionSE())
    ManagerFunctions.stopMusic("BGM")
    local nextScene = SceneMap.new()
    nextScene:setInst(GameInstance.new())
    GlobalSystem.setScene(nextScene)
end

function Scene:_onLoadCommand()
    AudioManager.playSound(SourceSystem.GetDecisionSE())
    self._windowCommand:setActive(false)
    self._windowSaveLoad:open()
end

---@param reason string
function Scene:_onSaveLoadClose(reason)
    if reason == "loaded" then
        return
    end
    self._windowCommand:setActive(true)
    self._windowCommand:requestKeyboardFocus()
end

---@param inst Source.GameInstance.GameInstance
---@diagnostic disable-next-line: unused
function Scene:_onSaveLoadLoaded(inst)
    local SceneMap = require("Source.Scenes.SceneMap")

    ManagerFunctions.stopMusic("BGM")
    local nextScene = SceneMap.new()
    nextScene:setInst(inst)
    GlobalSystem.setScene(nextScene)
end

function Scene:_onConfigCommand()
    AudioManager.playSound(SourceSystem.GetDecisionSE())
    if self._configWindow:isOpen() then
        self._configWindow:close()
    else
        self._configWindow:open()
        self._windowCommand:setActive(false)
    end
end

function Scene:_onConfigClose()
    self._windowCommand:setActive(true)
    self._windowCommand:requestKeyboardFocus()
end

---@diagnostic disable-next-line: unused
function Scene:_exitGame()
    AudioManager.playSound(SourceSystem.GetDecisionSE())
    GlobalSystem.exit()
end

return class(Scene, SceneBase)
